### PR TITLE
Properties in SimpleXMLElement may be null

### PR DESCRIPTION
**Pull request by orklah**
As discussed in phpstan/phpstan/issues/2784 this returns a BenevolentUnionType(SimpleXMLElement|null) on properties for SimpleXMLElement.

It then allows operations on BenevolentUnionType by overriding unionResults (the method implementing the logic for whether an operation is permitted). While UnionType allow an operation if it's allowed on every Type in the Union, BenevolenUnionType allow the operation if it's allowed on at least one Type in the Union.

I added a test on impossible-instanceof. Without the modification in SimpleXML, it returned
```
385: Instanceof between SimpleXMLElement and SimpleXMLElement will always evaluate to true.
```
and with the modification, it doesn't return anything.

As you said in phpstan/phpstan/issues/2784:
> It's completely valid use-case that people are sure about their XML structure and know some attribute isn't going to be null.

I believe some functions in SimpleXMLElement should return BenevolentUnionType as well, such as SimpleXMLElement::attributes, who currently return SimpleXMLElement|null. Unfortunately, I tried modifying the type to (SimpleXMLElement|null) in FunctionMap but it doesn't seem to be resolved into a BenevolentUnionType. This could be for a future PR.

### CODE REVIEW

**Comment by orklah** [comment]
I couldn't run phpcbf on my windows PHP 7.4 because of deprecations and fatal errors due to an old version of squizlabs\php_codesniffer. I will check on the CI here to fix potential issues

File: src/Type/BenevolentUnionType.php [hunk 1]
```diff
@@ -42,4 +42,13 @@ public function isAcceptedBy(Type $acceptingType, bool $strictTypes): TrinaryLog
 		return TrinaryLogic::createNo();
 	}
 
+	/**
+	 * @param callable(Type $type): TrinaryLogic $getResult
+	 * @return TrinaryLogic
+	 */
+	protected function unionResults(callable $getResult): TrinaryLogic
+	{
+		return TrinaryLogic::maxMin(...array_map($getResult, $this->getTypes()));
```
Comment (orklah) [inline src/Type/BenevolentUnionType.php]: I'm not certain maxMin was the best method to use here, but I wasn't sure what the Maybe case...

**Comment by orklah** [comment]
@ondrejmirtes I think this is ready to be reviewed. Please tell me if I made something wrong or if I should implement more tests or anything!

Thanks!

**Comment by ondrejmirtes** [comment]
Merged as: https://github.com/phpstan/phpstan-src/commit/fcfcbd5bf1ce4a2d304549e801723fe1970623eb

Sorry to keep you waiting!

**Comment by orklah** [comment]
Wow! I lost hope on this one a looong time ago :D

Thanks Ondrej!

**Pull request closed without merging**